pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,426 @@
+	// TODO: Add a compressor that intelligently squashes together updates.
+	keyLister := cache.KeyListerFunc(func() []string { return s.listKeys() })
+	serviceQueue := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, keyLister)
+	lw := cache.NewListWatchFromClient(s.kubeClient.(*client.Client), "services", api.NamespaceAll, fields.Everything())
+	cache.NewReflector(lw, &api.Service{}, serviceQueue, 0).Run()
+	// TODO: Add proper retries rather than just re-adding to the queue?
+	for {
+		newItem := serviceQueue.Pop()
+		deltas, ok := newItem.(cache.Deltas)
+		if !ok {
+			glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", newItem)
+		}
+		delta := deltas.Newest()
+		if delta == nil {
+			glog.Errorf("Received nil delta from watcher queue.")
+			continue
+		}
+		err := s.processDelta(delta)
+		if err != nil {
+			// Add the failed service back to the queue so we'll retry it.
+			glog.Error(err)
+			time.Sleep(5 * time.Second)
+			serviceQueue.AddIfNotPresent(deltas)
+			break
+		}
+	}
+}
+
+// Returns an error if processing the delta should be retried, otherwise returns nil.
+func (s *ServiceController) processDelta(delta *cache.Delta) error {
+	service, ok := delta.Object.(*api.Service)
+	if !ok {
+		// If the DeltaFIFO saw a key in our cache that it didn't know about, it
+		// can send a deletion with an unknown state. Grab the service from our
+		// cache for deleting.
+		key, ok := delta.Object.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Delta contained object that wasn't a service or a deleted key: %+v", delta)
+			return nil
+		}
+		service, ok = s.getService(key.Key)
+		if !ok {
+			glog.Errorf("Service %s not in cache even though the watcher thought it was. Ignoring the deletion.", key)
+			return nil
+		}
+		delta.Object = service
+	}
+	glog.V(2).Infof("Got new %s delta for service: %+v", delta.Type, service)
+
+	// TODO: Make this more parallel. The only things that need to serialized
+	// are changes to services with the same namespace and name.
+	// TODO: Handle added, updated, and sync differently?
+	switch delta.Type {
+	case cache.Added:
+		fallthrough
+	case cache.Updated:
+		fallthrough
+	case cache.Sync:
+		return s.createLoadBalancerIfNeeded(service)
+	case cache.Deleted:
+		return s.handleDelete(service)
+	default:
+		glog.Errorf("Unexpected delta type: %v", delta.Type)
+	}
+	return nil
+}
+
+func (s *ServiceController) createLoadBalancerIfNeeded(service *api.Service) error {
+	namespacedName, err := cache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		glog.Errorf("Couldn't generate namespaced name for service: %v", err)
+		return nil
+	}
+
+	cachedService, cached := s.getService(namespacedName)
+	if cached && !needsUpdate(cachedService, service) {
+		glog.Infof("LB already exists and doesn't need update for service %s", namespacedName)
+		return nil
+	}
+	if cached {
+		// If the service already exists but needs to be updated, delete it so that
+		// we can recreate it cleanly.
```

Comment:
That's a very fair point. Unfortunately, GCE forwarding rules can't be updated in place other than to point them at a different target pool.

However, we can break down the problem a bit. The reasons we may need to update an external load balancer (as you can determine from the needsUpdate function) are if (a) the public IPs changed, or (b) the ports changed.

There are a couple alternative ways to handle these that might be worth considering. For changing ports, we could:
1. Create a new forwarding rule pointed at the same target pool with the new ports, then update the REST object with the new IP address this would create, then delete the old one. The downside is that this would require changing the IP and tracking the names of the forwarding rules rather than knowing exactly what the name could be.
2. Always set the port range to be 1-65535, regardless of which ports the service actually uses. Then we can disregard changes to the port numbers.
3. Accept a bit of downtime, but at least keep the IP around. Reserve the IP the FW is using, then delete the old FW, then create the new one with the correct IP, then un-reserve the IP. However, reserving IPs is GCE-specific, so it'd have to go in the GCE cloud provider implementation of create and update load balancer.

For changing IP addresses, we could:
1. The same as option 1 above.
2. Put the IP address in the name of the load balancer, so that we could easily create the new one while the old one still exists while still having deterministic names. It'd be tricky to do this for new load balancers, though.

For now, accepting a bit of downtime on service update doesn't seem absurd, but we could try one of these other options as a future issue.

Really, it seems like IP addresses chosen by the cloud provider should be in the service's status, not the service's spec, since this code can't determine whether the IP address was user-specified and needs to be respected vs whether it was arbitrarily chosen by the cloud provider when the previous balancer was created. Do you or @thockin think this sort of change is feasible given the recent switch to v1beta3?